### ISSUE TITLE
arith exception when using storable vector fromList/toList on 0 byte storable values

### ISSUE THREAD

**Issue by cartazio** [haskell/vector#105]
this used to work (seemingly) in vector 0.10, or perhaps primitive 0.5, but something changed in the two or the underlying backing operations in the past year.

``` haskell
import Foreign.Storable
import Data.Vector.Storable as SV 
data Nil = Nil 
  deriving Show 


instance Storable (Nil) where
    {-#INLINE sizeOf#-}
    sizeOf = \ _ ->  0
    -- might want to boost the alignment, but ignore for now
    {-# INLINE alignment #-}
    alignment = \ _ -> 1 
    {-# INLINE peek #-}
    peek = \ _  -> return Nil
    {-# INLINE poke #-}
    poke = \ _  _-> return ()
    {-# INLINE pokeElemOff #-}
    {-# INLINE peekElemOff #-}
    peekElemOff = \ _ _  -> return Nil
    pokeElemOff = \ _ _ _  -> return ()

    peekByteOff  = \ _ _ -> return Nil
    pokeByteOff  = \ _ _ _ -> return ()
    {-# INLINE peekByteOff #-}
    {-# INLINE pokeByteOff #-}

main =  print $ SV.toList$  SV.fromList  [Nil]

{-# NONINLINE foo#-}
foo :: [Nil] -> SV.Vector Nil 
foo = SV.fromList 
```

**Comment by cartazio** [comment]
the included program should trip the issue @rwbarton @dolio  @ekmett

**Comment by ekmett** [comment]
I think we now divide out by the length somewhere when figuring out how to initialize memory for better determinism now. This seems to demand a branch in that logic.

**Comment by cartazio** [comment]
Oh yeah the zeroing thing!

**Comment by cartazio** [comment]
I'll throw together a fix and regression test this weekend, this seems like a pretty nasty (though hopefully rare for most folks) bug :)

**Comment by mihaimaruseac** [comment]
I think this is now closed by #168
